2.10.3.Final -> ServiceConfigurationError: SmallRyeConfigFactory: QuarkusConfigFactory not a subtype

### Describe the bug

When moving from 2.10.2.Final to 2.10.3.Final, I see that some of our tests fails with this error, see test output below.


https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/2.2E10.2E3.2EFinal.20-.3E.20ServiceConfigurationError.3A.20SmallRyeConfigFac

### Expected behavior

The test should run ok as in previous version 2.10.2.Final 

### Actual behavior

The test fails:
[ERROR] app.ArkivControllerMockTest.tst  Time elapsed: 0.772 s  <<< ERROR!
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
Caused by: java.lang.reflect.InvocationTargetException
Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype

### How to Reproduce?

run the provided test and observer the error
The test run will succeed if you revert back to: 2.10.2.Final 
The test will also succeed if the CircuitBreaker are removed from the interface
[bsf-online2.zip](https://github.com/quarkusio/quarkus/files/9195673/bsf-online2.zip)



### Output of `uname -a` or `ver`

Linux hty80028.sb1a.sparebank1.no 4.18.0-372.16.1.el8_6.x86_64 #1 SMP Tue Jun 28 03:02:21 EDT 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.3" 2022-04-19 OpenJDK Runtime Environment Temurin-17.0.3+7 (build 17.0.3+7) OpenJDK 64-Bit Server VM Temurin-17.0.3+7 (build 17.0.3+7, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

The shortened stack trace in this issue report doesn't show anything, but the full stacktrace in the Zulip discussion shows that the problem occurs at

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at io.quarkus.test.junit.QuarkusTestExtension.popMockContext(QuarkusTestExtension.java:675)
    at io.quarkus.test.junit.QuarkusTestExtension.afterEach(QuarkusTestExtension.java:510)
```

and the `popMock` method calls

```
Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
    at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:593)
    at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1244)
    at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
    at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
    at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
    at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:100)
    at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
    at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
    at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
    at io.smallrye.faulttolerance.config.CircuitBreakerConfigImpl.delayUnit(CircuitBreakerConfigImpl.java:180)
    at io.smallrye.faulttolerance.FaultToleranceInterceptor.prepareSyncStrategy(FaultToleranceInterceptor.java:365)
    at io.smallrye.faulttolerance.FaultToleranceInterceptor.lambda$syncFlow$2(FaultToleranceInterceptor.java:250)
    at io.smallrye.faulttolerance.internal.StrategyCache.lambda$getStrategy$0(StrategyCache.java:20)
    at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
    at io.smallrye.faulttolerance.internal.StrategyCache.getStrategy(StrategyCache.java:20)
    at io.smallrye.faulttolerance.FaultToleranceInterceptor.syncFlow(FaultToleranceInterceptor.java:250)
    at io.smallrye.faulttolerance.FaultToleranceInterceptor.intercept(FaultToleranceInterceptor.java:182)
    at io.smallrye.faulttolerance.FaultToleranceInterceptor_Bean.intercept(Unknown Source)
    at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
    at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:40)
    at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
    at sb1.fip.app.bsf.online.client.SkadeTilbudDocClient$$CDIWrapper_Subclass.clearMock(Unknown Source)
    at sb1.fip.app.bsf.online.client.SkadeTilbudDocClient$$CDIWrapper_ClientProxy.clearMock(Unknown Source)
    at io.quarkus.test.junit.MockSupport.popContext(MockSupport.java:28)
```

That is, during `@AfterEach`, we call `clearMock`, which gets intercepted (because the RestClient interface has a class-level fault tolerance annotation), and that is wrong.

The fix is simple: the 2 methods related to mocking on `RestClientReactiveCDIWrapperBase` are missing the `@NoClassInterceptors` annotation. I'll submit a PR, though I'll first take a look whether there's a simpler / more direct test than the reproducer attached.

--- a/extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientReactiveCDIWrapperBase.java
+++ b/extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientReactiveCDIWrapperBase.java
@@ -48,11 +48,13 @@ public abstract class RestClientReactiveCDIWrapperBase<T extends Closeable> impl
     }
 
     @Override
+    @NoClassInterceptors
     public void setMock(Object mock) {
         this.mock = mock;
     }
 
     @Override
+    @NoClassInterceptors
     public void clearMock() {
         this.mock = null;
     }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
@@ -160,14 +160,15 @@ final class Methods {
                 bytecodeTransformerConsumer, transformUnproxyableClasses,
                 new SubclassSkipPredicate(beanDeployment.getAssignabilityCheck()::isAssignableFrom,
                         beanDeployment.getBeanArchiveIndex()),
-                false);
+                false, new HashSet<>());
     }
 
-    static Set<MethodInfo> addInterceptedMethodCandidates(BeanDeployment beanDeployment, ClassInfo classInfo,
+    private static Set<MethodInfo> addInterceptedMethodCandidates(BeanDeployment beanDeployment, ClassInfo classInfo,
             ClassInfo originalClassInfo,
             Map<MethodKey, Set<AnnotationInstance>> candidates,
             Set<AnnotationInstance> classLevelBindings, Consumer<BytecodeTransformer> bytecodeTransformerConsumer,
-            boolean transformUnproxyableClasses, SubclassSkipPredicate skipPredicate, boolean ignoreMethodLevelBindings) {
+            boolean transformUnproxyableClasses, SubclassSkipPredicate skipPredicate, boolean ignoreMethodLevelBindings,
+            Set<MethodKey> noClassInterceptorsMethods) {
 
         Set<NameAndDescriptor> methodsFromWhichToRemoveFinal = new HashSet<>();
         Set<MethodInfo> finalMethodsFoundAndNotChanged = new HashSet<>();
@@ -175,7 +176,7 @@ final class Methods {
 
         for (MethodInfo method : classInfo.methods()) {
             Set<AnnotationInstance> merged = mergeBindings(beanDeployment, originalClassInfo, classLevelBindings,
-                    ignoreMethodLevelBindings, method);
+                    ignoreMethodLevelBindings, method, noClassInterceptorsMethods);
             if (merged.isEmpty() || skipPredicate.test(method)) {
                 continue;
             }
@@ -206,7 +207,7 @@ final class Methods {
                 finalMethodsFoundAndNotChanged
                         .addAll(addInterceptedMethodCandidates(beanDeployment, superClassInfo, classInfo, candidates,
                                 classLevelBindings, bytecodeTransformerConsumer, transformUnproxyableClasses, skipPredicate,
-                                ignoreMethodLevelBindings));
+                                ignoreMethodLevelBindings, noClassInterceptorsMethods));
             }
         }
 
@@ -216,16 +217,33 @@ final class Methods {
                 //interfaces can't have final methods
                 addInterceptedMethodCandidates(beanDeployment, interfaceInfo, originalClassInfo, candidates,
                         classLevelBindings, bytecodeTransformerConsumer, transformUnproxyableClasses,
-                        skipPredicate, true);
+                        skipPredicate, true, noClassInterceptorsMethods);
             }
         }
         return finalMethodsFoundAndNotChanged;
     }
 
     private static Set<AnnotationInstance> mergeBindings(BeanDeployment beanDeployment, ClassInfo classInfo,
-            Set<AnnotationInstance> classLevelBindings, boolean ignoreMethodLevelBindings, MethodInfo method) {
+            Set<AnnotationInstance> classLevelBindings, boolean ignoreMethodLevelBindings, MethodInfo method,
+            Set<MethodKey> noClassInterceptorsMethods) {
 
-        if (beanDeployment.getAnnotation(method, DotNames.NO_CLASS_INTERCEPTORS) != null) {
+        MethodKey key = new MethodKey(method);
+        if (beanDeployment.getAnnotation(method, DotNames.NO_CLASS_INTERCEPTORS) != null
+                || noClassInterceptorsMethods.contains(key)) {
+            // The set of methods with `@NoClassInterceptors` is shared in the traversal of class hierarchy, so once
+            // a method with the annotation is found, all subsequent occurences of the "same" method are treated
+            // as if they also had it. Given that we traverse classes bottom-up, this works as expected: presence
+            // of `@NoClassInterceptors` on subclass overrides absence on superclass, and absence on subclass overrides
+            // presence on superclass. We process interfaces after all superclasses, so presence/absence on a class
+            // overrides absence/presence on a `default` method inherited from an interface. That is also expected.
+            // (Note that we apply class-level interceptors to `default` methods, as an extension to the specification,
+            // so we have to care.) However, due to recursive nature of the traversal, interfaces implemented by
+            // superclasses are processed _before_ interfaces implemented by subclasses. This still leads to expected
+            // behavior when a `default` method declares `@NoClassInterceptors` and is inherited only once, because
+            // inheritance chain of each interface is also processed bottom-up. However, if a `default` method from
+            // an interface is inherited multiple times, `@NoClassInterceptors` declared on that method may behave
+            // weirdly. This is enough of a corner case that we'll leave it and solve it when it becomes problematic.
+            noClassInterceptorsMethods.add(key);
             classLevelBindings = Set.of();
         }
 
@@ -320,7 +338,7 @@ final class Methods {
         final String name;
         final List<DotName> params;
         final DotName returnType;
-        final MethodInfo method;
+        final MethodInfo method; // this is intentionally ignored for equals/hashCode
 
         public MethodKey(MethodInfo method) {
             this.method = Objects.requireNonNull(method, "Method must not be null");

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/InterceptedBean.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/InterceptedBean.java
@@ -5,7 +5,7 @@ import javax.enterprise.context.ApplicationScoped;
 
 @ApplicationScoped
 @ClassLevel
-public class InterceptedBean extends SuperclassWithInterceptor {
+public class InterceptedBean extends SuperclassWithInterceptor implements Superinterface {
     @MethodLevel
     @NoClassInterceptors
     public InterceptedBean() {

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/NoClassInterceptorsTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/NoClassInterceptorsTest.java
@@ -83,5 +83,17 @@ public class NoClassInterceptorsTest {
         assertEquals(4, ClassLevelInterceptor.AROUND_INVOKE_COUNTER);
         assertEquals(4, InheritedClassLevelInterceptor.AROUND_INVOKE_COUNTER);
         assertEquals(4, MethodLevelInterceptor.AROUND_INVOKE_COUNTER);
+
+        bean.defaultMethod();
+
+        assertEquals(4, ClassLevelInterceptor.AROUND_INVOKE_COUNTER);
+        assertEquals(4, InheritedClassLevelInterceptor.AROUND_INVOKE_COUNTER);
+        assertEquals(4, MethodLevelInterceptor.AROUND_INVOKE_COUNTER);
+
+        bean.inheritedDefaultMethod();
+
+        assertEquals(5, ClassLevelInterceptor.AROUND_INVOKE_COUNTER);
+        assertEquals(5, InheritedClassLevelInterceptor.AROUND_INVOKE_COUNTER);
+        assertEquals(4, MethodLevelInterceptor.AROUND_INVOKE_COUNTER);
     }
 }

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/SuperSuperinterface.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/SuperSuperinterface.java
@@ -1,0 +1,9 @@
+package io.quarkus.arc.test.interceptors.noclassinterceptors;
+
+import io.quarkus.arc.NoClassInterceptors;
+
+public interface SuperSuperinterface {
+    @NoClassInterceptors
+    default void inheritedDefaultMethod() {
+    }
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/Superinterface.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/noclassinterceptors/Superinterface.java
@@ -1,0 +1,22 @@
+package io.quarkus.arc.test.interceptors.noclassinterceptors;
+
+import io.quarkus.arc.NoClassInterceptors;
+
+public interface Superinterface extends SuperSuperinterface {
+    void inheritedClassLevelAndMethodLevel();
+
+    void inheritedMethodLevelOnly();
+
+    void inheritedNoInterceptors();
+
+    @MethodLevel // this should be ignored, only class-level interceptors apply to default methods
+    @NoClassInterceptors // and this makes sure that even class-level interceptors do not apply
+    default void defaultMethod() {
+    }
+
+    // `SuperSuperinterface.inheritedDefaultMethod` has `@NoClassInterceptors`, but this override
+    // does not, so class-level interceptors _should_ apply
+    @Override
+    default void inheritedDefaultMethod() {
+    }
+}

--- a/integration-tests/rest-client-reactive/src/main/java/io/quarkus/it/rest/client/main/ClientCallingResource.java
+++ b/integration-tests/rest-client-reactive/src/main/java/io/quarkus/it/rest/client/main/ClientCallingResource.java
@@ -155,13 +155,13 @@ public class ClientCallingResource {
         });
 
         router.route("/call-with-fault-tolerance-on-interface").blockingHandler(rc -> {
-            String exception = "";
+            String result = "";
             try {
-                faultToleranceOnInterfaceClient.hello();
+                result = faultToleranceOnInterfaceClient.hello();
             } catch (Exception e) {
-                exception = e.getClass().getSimpleName();
+                result = e.getClass().getSimpleName();
             }
-            rc.end(exception);
+            rc.end(result);
         });
 
         router.get("/with%20space").handler(rc -> rc.response().setStatusCode(200).end());

--- a/integration-tests/rest-client-reactive/src/test/java/io/quarkus/it/rest/client/InjectMockWithInterceptorTest.java
+++ b/integration-tests/rest-client-reactive/src/test/java/io/quarkus/it/rest/client/InjectMockWithInterceptorTest.java
@@ -1,0 +1,50 @@
+package io.quarkus.it.rest.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+import javax.inject.Inject;
+
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.it.rest.client.main.FaultToleranceOnInterfaceClient;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.mockito.InjectMock;
+import io.restassured.RestAssured;
+import io.smallrye.faulttolerance.api.CircuitBreakerMaintenance;
+
+@QuarkusTest
+public class InjectMockWithInterceptorTest {
+    @Inject
+    CircuitBreakerMaintenance cb;
+
+    @InjectMock
+    FaultToleranceOnInterfaceClient mock;
+
+    @BeforeEach
+    public void setUp() {
+        cb.resetAll();
+        when(mock.hello()).thenReturn("MockHello");
+    }
+
+    @AfterEach
+    public void tearDown() {
+        cb.resetAll();
+    }
+
+    @Test
+    void shouldMockClient() {
+        assertThat(mock.hello()).isEqualTo("MockHello");
+    }
+
+    @Test
+    void shouldMockClientInTheApp() {
+        RestAssured.with().post("/call-with-fault-tolerance-on-interface")
+                .then()
+                .statusCode(200)
+                .body(Matchers.equalTo("MockHello"));
+    }
+}

